File: CStoreManagement/app/src/main/java/com/c_store/zhiyazhang/cstoremanagement/view/activity/NCScrapActivity.java
package com.c_store.zhiyazhang.cstoremanagement.view.activity;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.c_store.zhiyazhang.cstoremanagement.R;
import com.c_store.zhiyazhang.cstoremanagement.bean.ScrapContractBean;
import com.c_store.zhiyazhang.cstoremanagement.presenter.scrap.BarCodeScrapAdapter;
import com.c_store.zhiyazhang.cstoremanagement.sql.MySql;
import com.c_store.zhiyazhang.cstoremanagement.sql.ScrapDao;
import com.c_store.zhiyazhang.cstoremanagement.utils.ConnectionDetector;
import com.c_store.zhiyazhang.cstoremanagement.utils.MyApplication;
import com.c_store.zhiyazhang.cstoremanagement.utils.MyLinearlayoutManager;
import com.c_store.zhiyazhang.cstoremanagement.utils.activity.MyActivity;
import com.c_store.zhiyazhang.cstoremanagement.utils.onclick.RecyclerAddLessClickListener;
import com.c_store.zhiyazhang.cstoremanagement.utils.recycler.ItemTouchHelperCallback;
import com.c_store.zhiyazhang.cstoremanagement.utils.socket.SocketUtil;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * Created by zhiya.zhang
 * on 2017/5/27 15:54.
 * 选完类后的具体商品
 * 不复杂所以也不用mvc了
 */

public class NCScrapActivity extends MyActivity {
    @BindView(R.id.my_toolbar)
    Toolbar myToolbar;
    @BindView(R.id.recyclerView2)
    RecyclerView recycler;
    BarCodeScrapAdapter adapter;
    Context mContext;
    @BindView(R.id.sure_btn)
    Button sureBtn;
    ScrapDao sd = new ScrapDao(this);
    @BindView(R.id.loading)
    LinearLayout loading;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        sureBtn.setVisibility(View.VISIBLE);
        initView();
        initDate();
    }

    private void initDate() {
        if (!ConnectionDetector.getConnectionDetector().isOnline()){
            Toast.makeText(NCScrapActivity.this,MyApplication.getContext().getResources().getString(R.string.noInternet),Toast.LENGTH_SHORT).show();
            return;
        }
        String ip = MyApplication.getIP();
        if (ip.equals(MyApplication.getContext().getResources().getString(R.string.notFindIP))) {
            Toast.makeText(NCScrapActivity.this,MyApplication.getContext().getResources().getString(R.string.notFindIP),Toast.LENGTH_SHORT).show();
            return;
        }
        showLoading();
        showLoading();
        SocketUtil.getSocketUtil(ip).inquire(MySql.getAllScrapByCategory(Integer.toString(getIntent().getIntExtra("categorynumber", 0))), new Handler() {
            @Override
            public void handleMessage(Message msg) {
                switch (msg.what) {
                    case 0:
                        if (((String) msg.obj).equals("") || ((String) msg.obj).equals("[]")) {
                            Toast.makeText(mContext, "无category", Toast.LENGTH_SHORT).show();
                            break;
                        }
                        ArrayList<ScrapContractBean> scbs = new Gson().fromJson((String) msg.obj, new TypeToken<List<ScrapContractBean>>() {
                        }.getType());
                        //检查已添加数据中是否有相同id的数据，有的话已添加数据覆盖当前数据
                        for (ScrapContractBean scb :
                                sd.getAllDate()) {
                            for (ScrapContractBean s :
                                    scbs) {
                                if (scb.getScrapId().equals(s.getScrapId())) {
                                    s.setNowMrkCount(scb.getNowMrkCount());
                                    s.setIsScrap(scb.getIsScrap());
                                    s.setIsNew(scb.getIsNew());
                                }
                            }
                        }
                        //得到adapter关联ItemTouchHelper和RecyclerView
                        adapter = new BarCodeScrapAdapter(scbs, 1);
                        adapter.setOnItemClickLitener(onItemClickListener);
                        recycler.setAdapter(adapter);
                        ItemTouchHelper.Callback callback = new ItemTouchHelperCallback(adapter);
                        ItemTouchHelper mItemTouchHelper = new ItemTouchHelper(callback);
                        mItemTouchHelper.attachToRecyclerView(recycler);
                        hideLoading();
                        break;
                    case 1:
                        Toast.makeText(mContext, (String) msg.obj, Toast.LENGTH_SHORT).show();
                        hideLoading();
                        break;
                    case 2:
                        Toast.makeText(mContext, (String) msg.obj, Toast.LENGTH_SHORT).show();
                        hideLoading();
                        break;
                }
            }
        });
    }

    private void initView() {
        mContext = NCScrapActivity.this;
        myToolbar.setTitle(getIntent().getStringExtra("categoryname"));
        setSupportActionBar(myToolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        recycler.setLayoutManager(new MyLinearlayoutManager(mContext, LinearLayoutManager.VERTICAL, false));
    }

    /**
     * recycler的监听
     */
    RecyclerAddLessClickListener onItemClickListener = new RecyclerAddLessClickListener() {
        @Override
        public void addItemOnClick(TextView tv, int position) {
            adapter.getList().get(position).setNowMrkCount(adapter.getList().get(position).getNowMrkCount() + 1);
            tv.setText(Integer.toString(adapter.getList().get(position).getNowMrkCount()));
        }

        @Override
        public void lessItemOnClick(TextView tv, int position) {
            if (adapter.getList().get(position).getNowMrkCount() != 0) {
                adapter.getList().get(position).setNowMrkCount(adapter.getList().get(position).getNowMrkCount() - 1);
                tv.setText(Integer.toString(adapter.getList().get(position).getNowMrkCount()));
            }
        }
    };

    Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case 528:
                    hideLoading();
                    finish();
                    break;
            }
        }
    };

    Thread mThread = new Thread(new Runnable() {
        @Override
        public void run() {
            try {
                Looper.prepare();
                ArrayList<ScrapContractBean> insertList = new ArrayList<ScrapContractBean>();
                ArrayList<ScrapContractBean> updateList = new ArrayList<ScrapContractBean>();

                for (ScrapContractBean scb :
                        adapter.getList()) {
                    //要修改过的，要不是已提交过的
                    if (scb.getNowMrkCount() > 0 && scb.getIsScrap() == 0) {
                        //如果是新的就创建，如果是老的就去修改
                        if (scb.getIsNew() != 2) {
                            insertList.add(scb);
                        } else {
                            updateList.add(scb);
                        }
                    }
                }
                if (insertList.size() != 0 || updateList.size() != 0) {
                    if (insertList.size() != 0) {
                        sd.editSQL(insertList, "insert");
                    }
                    if (updateList.size() != 0) {
                        sd.editSQL(updateList, "update");
                    }
                    Message msg = new Message();
                    msg.what = 528;
                    mHandler.sendMessage(msg);
                } else {
                    Toast.makeText(mContext, "无要保存的数据", Toast.LENGTH_SHORT).show();
                }
            } catch (Exception e) {
                Log.e("NCScrapActivity", "", e);
            } finally {
                Looper.loop();
            }

        }
    });

    @OnClick(R.id.sure_btn)
    void sureBtn() {
        showLoading();
        mThread.start();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_scrap_nobarcode;
    }

    private void showLoading() {
        loading.setVisibility(View.VISIBLE);
    }

    private void hideLoading() {
        loading.setVisibility(View.GONE);
    }
}

File: CStoreManagement/app/src/main/java/com/cstore/zhiyazhang/cstoremanagement/view/scrap/NCScrapActivity.kt
/*
package com.cstore.zhiyazhang.cstoremanagement.view.scrap

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.helper.ItemTouchHelper
import android.util.Log
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import com.cstore.zhiyazhang.cstoremanagement.R
import com.cstore.zhiyazhang.cstoremanagement.bean.ScrapCategoryBean
import com.cstore.zhiyazhang.cstoremanagement.bean.ScrapContractBean
import com.cstore.zhiyazhang.cstoremanagement.presenter.scrap.BarCodeScrapAdapter
import com.cstore.zhiyazhang.cstoremanagement.sql.MySql
import com.cstore.zhiyazhang.cstoremanagement.sql.ScrapDao
import com.cstore.zhiyazhang.cstoremanagement.utils.ConnectionDetector
import com.cstore.zhiyazhang.cstoremanagement.utils.MyActivity
import com.cstore.zhiyazhang.cstoremanagement.utils.recycler.MyLinearlayoutManager
import com.cstore.zhiyazhang.cstoremanagement.utils.recycler.RecyclerOnTouch
import com.cstore.zhiyazhang.cstoremanagement.utils.socket.SocketUtil
import com.cstore.zhiyazhang.cstoremanagement.view.interfaceview.GenericView
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.zhiyazhang.mykotlinapplication.utils.MyApplication
import com.zhiyazhang.mykotlinapplication.utils.recycler.ItemTouchHelperCallback
import com.zhiyazhang.mykotlinapplication.utils.recycler.onMoveAndSwipedListener
import kotlinx.android.synthetic.main.activity_scrap_nobarcode.*
import kotlinx.android.synthetic.main.loading_layout.*
import kotlinx.android.synthetic.main.sure_btn.*
import kotlinx.android.synthetic.main.toolbar_layout.*

*/
/**
 * Created by zhiya.zhang
 * on 2017/8/8 10:10.
 *//*

class NCScrapActivity(override val layoutId: Int = R.layout.activity_scrap_nobarcode) : MyActivity(), GenericView {

    val sd = ScrapDao(this)
    var adapter: BarCodeScrapAdapter? = null
    val layoutManager = MyLinearlayoutManager(this, LinearLayoutManager.VERTICAL, false)
    var scrapCategory: ScrapCategoryBean? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sure_btn.visibility = View.VISIBLE
        scrapCategory = intent.getSerializableExtra("scrapCategory") as ScrapCategoryBean
        initView()
        initData()
        initClick()
    }

    private fun initView() {
        my_toolbar.title = scrapCategory!!.categoryName
        my_toolbar.setNavigationIcon(R.drawable.ic_action_back)
        setSupportActionBar(my_toolbar)
        recyclerView2.layoutManager = layoutManager
    }

    private fun initData() {
        showLoading()
        if (!ConnectionDetector.getConnectionDetector().isOnline) {
            showPrompt(getString(R.string.noInternet))
            hideLoading()
            return
        }
        val ip = MyApplication.getIP()
        if (ip == MyApplication.instance().getString(R.string.notFindIP)) {
            showPrompt(ip)
            hideLoading()
            return
        }
        SocketUtil.getSocketUtil(ip).inquire(MySql.getAllScrapByCategory(scrapCategory!!.categoryId), 20, @SuppressLint("HandlerLeak")
        object : Handler() {
            override fun handleMessage(msg: Message) {
                when (msg.what) {
                    0 -> {
                        if (msg.obj.toString() == "" || msg.obj.toString() == "[]") {
                            showPrompt(getString(R.string.noMessage))
                            hideLoading()
                            return
                        }
                        //检查已添加数据中是否有相同id的数据，有的话已添加数据覆盖当前数据
                        val scbs = Gson().fromJson<ArrayList<ScrapContractBean>>(msg.obj.toString(), object : TypeToken<ArrayList<ScrapContractBean>>() {}.type)
                        for (scb in sd.allDate) {
                            for (s in scbs) {
                                if (scb.scrapId == s.scrapId) {
                                    s.mrkCount = scb.mrkCount
                                    s.isScrap = scb.isScrap
                                    s.isNew = scb.isNew
                                }
                            }
                        }

                        adapter = BarCodeScrapAdapter(scbs, onClick, 1)
                        recyclerView2.adapter = adapter
                        val callback = ItemTouchHelperCallback(adapter as onMoveAndSwipedListener)
                        val mItemTouchHelper = ItemTouchHelper(callback)
                        mItemTouchHelper.attachToRecyclerView(recyclerView2)
                        hideLoading()
                    }
                    else -> {
                        showPrompt(msg.obj.toString())
                        hideLoading()
                    }
                }
            }
        })
    }

    val onClick = object : RecyclerOnTouch {
        override fun <T> onClickImage(objects: T, position: Int) {}

        override fun <T> onTouchAddListener(objects: T, event: MotionEvent, position: Int) {
            val adapterView: BarCodeScrapAdapter.ViewHolder = recyclerView2.findViewHolderForAdapterPosition(position) as BarCodeScrapAdapter.ViewHolder
            onTouchChange(1, event.action, adapterView, position)
        }

        override fun <T> onTouchLessListener(objects: T, event: MotionEvent, position: Int) {
            val adapterView: BarCodeScrapAdapter.ViewHolder = recyclerView2.findViewHolderForAdapterPosition(position) as BarCodeScrapAdapter.ViewHolder
            onTouchChange(0, event.action, adapterView, position)
        }
    }

    private fun initClick() {
        sure_btn.setOnClickListener {
            showLoading()
            mThread.start()
        }
    }

    private val mHandler = @SuppressLint("HandlerLeak")
    object : Handler() {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                528 -> {
                    hideLoading()
                    finish()
                }
            }
        }
    }
    private val mThread = Thread(Runnable {
        kotlin.run {
            try {
                Looper.prepare()
                val insertList = ArrayList<ScrapContractBean>()
                val updateList = ArrayList<ScrapContractBean>()
                adapter!!.data
                        .filter { it.mrkCount > 0 && it.isScrap == 0 }
                        .forEach { if (it.isNew != 2) insertList.add(it) else updateList.add(it) }
                if (insertList.size != 0 || updateList.size != 0) {
                    if (insertList.size != 0) sd.editSQL(insertList, "insert")
                    if (updateList.size != 0) sd.editSQL(updateList, "update")
                    val msg = Message()
                    msg.what = 528
                    mHandler.sendMessage(msg)
                } else showPrompt(getString(R.string.noSaveMessage))
            } catch (e: Exception) {
                hideLoading()
                showPrompt(e.message.toString())
                Log.e("NCScrapActivity", "", e)
            } finally {
                Looper.loop()
            }
        }
    })

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finishAfterTransition()
        }
        return true
    }

    override fun <T> requestSuccess(objects: T) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showLoading() {
        loading.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        loading.visibility = View.GONE
    }

    override fun <T> showView(adapter: T) {
    }

    override fun errorDealWith() {
    }

    var mt: Thread? = null
    var pt: Thread? = null
    var hd: Handler? = null
    var isOnLongClick = false
    private fun onTouchChange(addLess: Int, action: Int, view: BarCodeScrapAdapter.ViewHolder, position: Int) {
        hd = @SuppressLint("HandlerLeak")
        object : Handler() {
            override fun handleMessage(msg: Message) {
                when (msg.what) {
                    0 -> if (view.less.isEnabled) lessCount(view, position)
                    1 -> if (view.add.isEnabled) addCount(view, position)
                }
            }
        }
        if (addLess == 0) {//less
            when (action) {
                MotionEvent.ACTION_DOWN -> {
                    mt = object : Thread(Runnable {
                        while (isOnLongClick) {
                            Thread.sleep(200)
                            hd!!.sendEmptyMessage(0)
                        }
                    }) {}
                    isOnLongClick = true
                    mt!!.start()
                }
                MotionEvent.ACTION_UP -> {
                    if (mt != null)
                        isOnLongClick = false
                    mt = null
                }
                MotionEvent.ACTION_MOVE -> {
                    if (mt != null && Integer.parseInt(view.scrapCount.text.toString()) != 0)
                        isOnLongClick = true
                }
            }
        } else {//add
            when (action) {
                MotionEvent.ACTION_DOWN -> {
                    pt = object : Thread(Runnable {
                        while (isOnLongClick) {
                            Thread.sleep(200)
                            hd!!.sendEmptyMessage(1)
                        }
                    }) {}
                    isOnLongClick = true
                    pt!!.start()
                }
                MotionEvent.ACTION_UP -> {
                    if (pt != null) isOnLongClick = false
                    pt = null
                }

                MotionEvent.ACTION_MOVE -> {
                    if (pt != null) isOnLongClick = true
                }
            }
        }
    }

    private fun addCount(view: BarCodeScrapAdapter.ViewHolder, position: Int) {
        if (adapter!!.data[position].mrkCount < 999) {
            adapter!!.data[position].mrkCount++
            view.scrapCount.text = adapter!!.data[position].mrkCount.toString()
        }
    }

    private fun lessCount(view: BarCodeScrapAdapter.ViewHolder, position: Int) {
        if (adapter!!.data[position].mrkCount > 0) {
            adapter!!.data[position].mrkCount--
            view.scrapCount.text = adapter!!.data[position].mrkCount.toString()
        }
    }
}*/
